[WIP,DNM] seastore: Integration work for RBM and CBJournal

myoungwon opened the pull request:
Signed-off-by: Myoungwon Oh <myoungwon.oh@samsung.com>


<!--
Thank you for opening a pull request!  Here are some tips on creating
a well formatted contribution.

Please give your pull request a title like "[component]: [short description]"

This is the format for commit messages:

"""
[component]: [short description]

[A longer multiline description]

Fixes: [ticket URL on tracker.ceph.com, create one if necessary]
Signed-off-by: [Your Name] <[your email]>
"""

The Signed-off-by line is important, and it is your certification that
your contributions satisfy the Developers Certificate or Origin.  For
more detail, see SubmittingPatches.rst.

The component is the short name of a major daemon or subsystem,
something like "mon", "osd", "mds", "rbd, "rgw", etc. For ceph-mgr modules,
give the component as "mgr/<module name>" rather than a path into pybind.

For more examples, simply use "git log" and look at some historical commits.

This was just a quick overview.  More information for contributors is available here:
https://raw.githubusercontent.com/ceph/ceph/master/SubmittingPatches.rst

-->
## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

athanatos commented:
Just a heads up: this PR is too big.  Before we start really reviewing this, I think you should be able to create several smaller PRs which are easier to digest.

myoungwon commented:
Ok.

athanatos commented:
Ah, nvm about my previous comment, I misunderstood the changes (though a focused PR will be easier to review and merge).  Btw, have you had a look at https://github.com/ceph/ceph/pull/41803?  I think there's some overlap with the ExtentAllocator concept here.

myoungwon commented:
We intend to address the following question on this PR.  First, extent allocator to make general interfaces—-for rbm and segment manager—-used by cache, transaction manager. Second, hybrid transaction procedure if rbm is used. Third, general journal interface to support SegmentJournal(existing journal) and CircularBoundedJournal (CBJournal). Updated commits are the first part. Before making further progress, we’d like to get feedback on this.

1. As your comment, we thought that extent allocator is needed to provide an abstraction for underlying allocators such as segment manager and rbm in order to add hybrid tansaction procedure—-TransactionManager(tm) hold segment manager(sm) or rbm, so it can provide different transaction processing depending on the type of allocator. But, segment journal and segment cleaner, both of which are intended only for journal, does not need to have general interfaces.
2. Regarding paddr_t, we probably need a unified address to cover both managers. I think existing paddr_t can work if segment filed is used as block id and offset is used as block’s offset in case rbm and cbjournal are used. What do you think?


Regarding extentplacementmanager PR (https://github.com/ceph/ceph/pull/41803), it seems that the purpose of it is to move the location of extent to another place. Therefore, is there any overlap with this PR? There is extent allocator in https://github.com/ceph/ceph/pull/41803 , but this is for alloc_extent(). We can re-arrange commits based on that changes.

xxhdx1985126 commented:
> 1. As your comment, we thought that extent allocator is needed to provide an abstraction for underlying allocators such as segment manager and rbm in order to add hybrid tansaction procedure—-TransactionManager(tm) hold segment manager(sm) or rbm, so it can provide different transaction processing depending on the type of allocator. But, segment journal and segment cleaner, both of which are intended only for journal, does not need to have general interfaces.

Could you elaborate a little bit about "hybrid transaction"?

xxhdx1985126 commented:
> Regarding extentplacementmanager PR (#41803), it seems that the purpose of it is to move the location of extent to another place. Therefore, is there any overlap with this PR? There is extent allocator in #41803 , but this is for alloc_extent(). We can re-arrange commits based on that changes.

I think there's naming issue in #41803, the "SegmentedRewriter" should be "SegmentedOolWriter" to be accurate, as it is used to write extents to other addresses than the journal, which is not necessarily a part of the extent rewrite.

cyx1231st commented:
> First, extent allocator to make general interfaces—-for rbm and segment manager—-used by cache, transaction manager.

Yeah, I have a similar thought in the comment https://github.com/ceph/ceph/pull/41803#pullrequestreview-687986018, in which I mentioned `SegmentedJournalAllocator` to abstract the existing logic of allocating inline extents from segmented jounal. IIUC RBM can leverage this general allocator to implement its own (whether out-of-line or inline). And PMem also needs an allocator. Then the PlacementManager can be the decision maker of extent placements by choossing an appropriate allocator (with general allocate interface, not just ool specific) according to the cold/hot hints and the configured devices.

> 1. ... But, segment journal and segment cleaner, both of which are intended only for journal, does not need to have general interfaces.

Not sure, can RBM journal and Segmented journal to share the same interface to be called by upper-layer components?

I think Cleaner can be a general concept for internal transactions to rearrage free-space per device. RBM might need a similar component to do defragmentation, right? Generalized cleaners may help to arrange all the cleaning efforts together according to dynamic workloads.

> 2. Regarding paddr_t, we probably need a unified address to cover both managers. I think existing paddr_t can work if segment filed is used as block id and offset is used as block’s offset in case rbm and cbjournal are used. What do you think?

Our current paddr is 64 bits, for example, 4 bit device-id, 28 bit segment-id and 32 bit segment-offset. RBM paddr should be the same overall width in order to be processed generally. The internal paddr layout can be different, for example, 4 bit device-id, 48 bit block-id and 12 bit block offset for 4KB blocks.

I'm still not sure the exact reason why 64-bit is not enough and need to extend to 128-bit (https://github.com/ceph/ceph/pull/41803#discussion_r665908885).

myoungwon commented:
@xxhdx1985126 I meant hybrid updates. I added the following code to show my thoughts.

```
TransactionManager::submit_transaction_direct_ret
TransactionManager::submit_transaction_hybrid(
  TransactionRef t)
{
  LOG_PREFIX(TransactionManager::submit_transaction_hybrid);
  auto &tref = *t;
  auto record = cache->try_construct_record(tref);
  if (!record) {
    DEBUGT("conflict detected, returning eagain.", tref);
    return crimson::ct_error::eagain::make();
  }

  DEBUGT("about to submit to rbm", tref);

  /*
   * After try_construct_record(), record_t contains both extents and deltas.
   * Among them, extents presumably contains large writes, so rbm handles such
   * extents before journaling records.
   *
   * rbm will write records if transaction includes large writes,
   * otherwise it does nothing.
   *
   * Once the writes to rbm is complete, other small writes including metadata
   * updates will be appended to journal (CircularBoundedJournal here).
   */
  return extent_allocator.submit_record(*record, tref.handle
  ).safe_then([this, FNAME, record, &tref]() mutable {
    vector<pair<paddr_t, bufferlist>> to_rbm;
    // The following code is added to store items which will be written to rbm
    // before submit_record.
    // This will be deleted after CircularBoundedJournal is in place.
    for (const auto &i: record->extents) {
      if (i.s_type == store_types_t::CBJOURNAL) {
        to_rbm.push_back(make_pair(i.rb_addr, i.bl));
      }
    }
    for (const auto &i: record->deltas) {
      if (i.s_type == store_types_t::CBJOURNAL) {
        to_rbm.push_back(make_pair(i.rb_addr, i.bl));
      }
    }
    return journal->submit_record(std::move(*record), tref.handle
    ).safe_then([this, FNAME, &tref, &to_rbm](auto p) mutable {
      auto [addr, journal_seq] = p;
      cache->complete_commit(tref, addr, journal_seq, nullptr);
      return lba_manager->complete_transaction(tref).safe_then(
        [&tref, this, &to_rbm] {
        /*
         * TODO: move data from journal to rbm
         * Due to lack of implemenation for CircularBoundedJournal,
         * we just write records to rbm directly here.
         */
        return crimson::do_for_each(to_rbm.begin(), to_rbm.end(),
          [&tref, this] (auto &val) {
          //auto addr = paddr_t{0, (segment_off_t)val.first.paddr};
          // TODO: need to find CacheExtentRef via rb_addr_t, not paddr_t
          return cache->get_extent_if_cached(tref, val.first
          ).then([&tref, &val, this](CachedExtentRef extent) mutable {
            ceph_assert(extent);
            auto ref = extent->cast<LogicalCachedExtent>();
            paddr_t rbm_addr = ref->get_rbm_addr();
            auto bptr = 
              bufferptr(ceph::buffer::create_page_aligned(val.second.length()));
            auto iter = val.second.cbegin();
            iter.copy(val.second.length(), bptr.c_str());
            auto rbm_abs_addr = rbm_addr.segment * extent_allocator.get_block_size() +
                            rbm_addr.offset;
            return extent_allocator.write(rbm_abs_addr, bptr
            ).safe_then([&tref, &ref, extent, rbm_addr, this]() {
              // set rbm address
              extent->set_rbm_addr(rbm_addr);
              /*
               * What we should do here is only updating physical address.
               * But, rewrite_extent creates another copy of CachedExtent.
               * How about exporting update_mapping to public to avoid
               * unnecessary copy overhead?
               */
              return lba_manager->rewrite_extent(tref, extent
              ).safe_then([]() {
                return SegmentManager::release_ertr::now();
              });
            });
          });
        });
      });
    }).safe_then([&tref] {
      return tref.handle.complete();
    }).handle_error(
      submit_transaction_ertr::pass_further{},
      crimson::ct_error::all_same_way([](auto e) {
        ceph_assert(0 == "Hit error submitting to journal");
      }));
  }).handle_error(
    submit_transaction_ertr::pass_further{},
    crimson::ct_error::all_same_way([](auto e) {
      ceph_assert(0 == "Hit error submitting to rbm");
    }));
}


```

xxhdx1985126 commented:
> @xxhdx1985126 I meant hybrid updates. I added the following code to show my though

I think this is overlapped with #41803, which is also trying to address the issue, I think maybe @cyx1231st 's way is an apprach to unify both of our work.

On the other hand, at the beginning when implementing #41803, I tried exactly the way shown in your code. But it turns out that this way may need a lot of auxiliary work. Because, extents submitted through "`extent_allocator.submit_record(*record, tref.handle)`" have to be referenced by internal nodes of the lba tree the modifications to which have to go into the same transaction as well, and when segment manager is used as the storage backend, the address of these extents can only be determined at the time of actual writes at which time we have passed the point to modify the internal nodes of lba trees. This is why #41803 works the way it is now. But, I'm assuming segment managers may be used as a backend for ool extents, if this is not the case, then all of the above analysis is pointless.

myoungwon commented:
> > First, extent allocator to make general interfaces—-for rbm and segment manager—-used by cache, transaction manager.
>
> Yeah, I have a similar thought in the comment [#41803 (review)](https://github.com/ceph/ceph/pull/41803#pullrequestreview-687986018), in which I mentioned `SegmentedJournalAllocator` to abstract the existing logic of allocating inline extents from segmented jounal. IIUC RBM can leverage this general allocator to implement its own (whether out-of-line or inline). And PMem also needs an allocator. Then the PlacementManager can be the decision maker of extent placements by choossing an appropriate allocator (with general allocate interface, not just ool specific) according to the cold/hot hints and the configured devices.

I agree.

>
> > 1. ... But, segment journal and segment cleaner, both of which are intended only for journal, does not need to have general interfaces.
>
> Not sure, can RBM journal and Segmented journal to share the same interface to be called by upper-layer components?
>

Yes, general Journal interface used by upper-layer components (work to do) is needed to handle both CircularBoundedJournal and SegmentJournal. 
I mean, since existing journal (segment journal) originally is designed only for SegmentManager, So, do we need to use extent allocator (wrapper for both SM and RBM) in the SegmentJournal, which is not shared components unlike cache and transaction manager. This makes extent allocator more complex. For example, SegmentJournal(SegmentManger& manager); needs to be SegmentJournal(ExtentAllocator& extent_allocator); ? 

> I think Cleaner can be a general concept for internal transactions to rearrage free-space per device. RBM might need a similar component to do defragmentation, right? Generalized cleaners may help to arrange all the cleaning efforts together according to dynamic workloads.
>
Right.
> > 1. Regarding paddr_t, we probably need a unified address to cover both managers. I think existing paddr_t can work if segment filed is used as block id and offset is used as block’s offset in case rbm and cbjournal are used. What do you think?
>
> Our current paddr is 64 bits, for example, 4 bit device-id, 28 bit segment-id and 32 bit segment-offset. RBM paddr should be the same overall width in order to be processed generally. The internal paddr layout can be different, for example, 4 bit device-id, 48 bit block-id and 12 bit block offset for 4KB blocks.
>
> I'm still not sure the exact reason why 64-bit is not enough and need to extend to 128-bit ([#41803 (comment)](https://github.com/ceph/ceph/pull/41803#discussion_r665908885)).

I was worried about address space limit due to the use of 32bit. If we can use 48 bit as block-id, it is reasonable to use current paddr without making changes.

myoungwon commented:
@xxhdx1985126 Can you explain which part is overlapped with #41803?  With RBM, the physical address can be determined before submitting transaction unlike SegmentManager. And, the two manager have different write procedures. How can #41803 handle write requests in the case of both small and large write in terms of rbm?  It seems that #41803 tried to only address how to write extent to other address.  

Also, how about using different submit_transaction code depending on the type of underlying allocator because the two write procedures are completely different.

xxhdx1985126 commented:
> @xxhdx1985126 Can you explain which part is overlapped with #41803? With RBM, the physical address can be determined before submitting transaction unlike SegmentManager. And, the two manager have different write procedures. How can #41803 handle write requests in the case of both small and large write in terms of rbm? It seems that #41803 tried to only address how to write extent to other address.

Sorry that I caused this confusion. Actually, I didn't mean the actual work is overlapped, I think only part of the concepts seems to be overlapping. I think there may be multiple types of devices are running underneath the seastore, all of which may be used to hold large writes or full extents, so there has to be a module to decide where to put those large writes. And #41803 is supposed to be that module, although we haven't inject any decision making algorithms into it yet. Right now, it can only support segment manager, but we've implemented the interfaces for futures device-specific allocators, and I think it's general enough to support rbm, just like what @cyx1231st said [here](https://github.com/ceph/ceph/pull/42270#issuecomment-881258183).

In terms of writing extents/large writes to other addresses than the journal, I think we are overlapped conceptually.

> 
> Also, how about using different submit_transaction code depending on the type of underlying allocator because the two write procedures are completely different.

I think this is fine. For segment manager backed devices, we use SegmentedOolAllocator and do large writes before constructing transactions; for rbm, we use RBMOolAllocator and first construct the transaction, then do large writes, and finally commit the transaction. What do you think about this?

xxhdx1985126 commented:
@myoungwon I think whether we are overlapped is dependent how we define hybrid transaction. If we define a hybrid transaction to be a journal record plus out-of-line large writes, we are overlapped conceptually on writing out-of-line large writes; otherwise, we are not.

myoungwon commented:
Let me describe what I understand until now. Please correct me.
There is ExtentPlacementManager(EPM). Its goal is to make a decision regarding where to put large write---the destination will be CircularBoundedJournal(CBJournal), SegmentManager(SM), RBM, or somewhere.
Using cached_extent via alloc_extent(*OolAllocator), EPM writes records depending on the decision, then updates update LBA after the write is complete except for the decision made before the write (in the case of RBM). At here, our work, ExtentAllocator which wraps existing both SM and RBM can co-work with EPM---EPM physically allocate extent via ExtentAllocator. Am I right?  Also, even though out-of-place(SegmentJournal) and in-place(CBJournal and RBM) are different write procedures. EPM is going to make a general interface for both cases. right?

I think, in terms of CBJournal and RBM, I need to know how the transaction works because of allocation management, partition, and update strategy between CBJournal and RBM on the RBM side. So, did you mean I need to update such work (transaction, RBMOolAllocator) based on EPM?

myoungwon commented:
> I think this is fine. For segment manager backed devices, we use SegmentedOolAllocator and do large writes before constructing transactions; for rbm, we use RBMOolAllocator and first construct the transaction, then do large writes, and finally commit the transaction. What do you think about this?

RBMOolAllocator will call both cache and RBM.alloc_extent? That is what I expected.


It seems that we need to clarify what hybrid transaction is.
In our design (RBM + CBJournal), small data---for instance, < 8KB---with metadata is appended CBJournal first,
then it is moved from CBjournal to RBM (in-place manner) including metadata.
Otherwise, if data is large--- > 8KB---, the data is written to RBM directly and the metadata is appended to CBJournal, 
then only the updated metadata is moved from CBJournal to RBM later.
So, depending on the data size, the transaction contains updates for either CBJournal and RBM or only CBJournal. 
This is different from the existing transaction procedure in seastore, so that's why I suggest another transaction code.
So, I guess this is an independent concept, am I wrong? what do you think?

xxhdx1985126 commented:
> > I think this is fine. For segment manager backed devices, we use SegmentedOolAllocator and do large writes before constructing transactions; for rbm, we use RBMOolAllocator and first construct the transaction, then do large writes, and finally commit the transaction. What do you think about this?
> 
> RBMOolAllocator will call both cache and RBM.alloc_extent? That is what I expected.

Yes, this is right.

> 
> It seems that we need to clarify what hybrid transaction is.
> In our design (RBM + CBJournal), small data---for instance, < 8KB---with metadata is appended CBJournal first,
> then it is moved from CBjournal to RBM (in-place manner) including metadata.
> Otherwise, if data is large--- > 8KB---, the data is written to RBM directly and the metadata is appended to CBJournal,
> then only the updated metadata is moved from CBJournal to RBM later.
> So, depending on the data size, the transaction contains updates for either CBJournal and RBM or only CBJournal.
> This is different from the existing transaction procedure in seastore, so that's why I suggest another transaction code.

Yes, I think this is where we go separate paths. Currently, I think that these large writes(> 8KB) are also ool writes. But I can accept your point, too:-)

> So, I guess this is an independent concept, am I wrong? what do you think?


I'm ok with either of the two definitions. It appears I'm not a man with hard ideas, hah:-). However, with regard to Journaling, I'm not quite sure whether it is appropriate for journal writes to go through EPM. I think the whole point of EPM is about choosing the right address for extents/large writes, whether it is on the same device as the journal or not. But I don't see there's any other choice to make when it comes to journal writes, I think journal writes should just go straight to the address pointed out by the journal pointer, so it looks to me that journal writes don't need to go through EPM. But this is only my thoughts, and it appears to conflict with that of @cyx1231st. I think we may need a thorough discussion about this point. @cyx1231st @tchaikov @athanatos @liu-chunmei :-)

myoungwon commented:
@athanatos  Before continuing this work, It would be better to get feedback on these commits I have updated.

1. I/O path: To make RBM be compatible with current seastore, I think that new extents including small and large block size need to be directly written to RBM, and all deltas excluding large data block need to be appended to CircularBoundedJournal(CBJournal). Then,
after *Cleaner---not implemented yet---reads records on CBJournal, it writes those records in RBM.

2. *Cleaner: This is needed to apply deltas from CBJournal to RBM, but this is likely to be the next PR.

3. paddr_t: current paddr_t has segment and offset fields. But, RBM probably needs to have block_id (block aligned) which is larger than segment, and block_offset which is less than offset.  So, I have adjusted the size of segment and offset field.

The followings are concerns I currently have.

1. Metadata section in RBM:
As you know, in RBM, data can be stored randomly across the entire RBM. Therefore, the existing replay mechanism---replaying records by reading record_t sequentially to load *nodes and data---does not work with RBM.
So, I think the metadata section is necessary in case of RBM.

2. Raw disk layout:
Currently, both CBjournal and RBM use record_t to store data (extents and deltas). This is reasonable in the case of CBjournal, but I’m not sure this if RBM is used due to its design characteristic (in-place)---record_t requires additional header when storing data. It seems that we may consider a conventional disk layout for in-place---ondoe contains extents, which indicates where the data is located---like ext3 in the future.


Any comments?

athanatos commented:
I'll try to have a look at this this week or next after https://github.com/ceph/ceph/pull/43181 merges. 

> @athanatos Before continuing this work, It would be better to get feedback on these commits I have updated.
> 
>     1. I/O path: To make RBM be compatible with current seastore, I think that new extents including small and large block size need to be directly written to RBM, and all deltas excluding large data block need to be appended to CircularBoundedJournal(CBJournal). Then,
>        after *Cleaner---not implemented yet---reads records on CBJournal, it writes those records in RBM.

I'd like the ExtentPlacementManager out-of-line extent support to be the machinery for this.  Have you looked into that yet?

> 
>     2. *Cleaner: This is needed to apply deltas from CBJournal to RBM, but this is likely to be the next PR.
> 

I think this probably needs to be a function of the segment cleaner as it's pretty similar to journal trimming with the existing implementation.

>     3. paddr_t: current paddr_t has segment and offset fields. But, RBM probably needs to have block_id (block aligned) which is larger than segment, and block_offset which is less than offset.  So, I have adjusted the size of segment and offset field.
> 

How large is a block?  Wouldn't you basically just want a byte offset with an assumption that all extents will be block aligned?

> 
> The followings are concerns I currently have.
> 
>     1. Metadata section in RBM:
>        As you know, in RBM, data can be stored randomly across the entire RBM. Therefore, the existing replay mechanism---replaying records by reading record_t sequentially to load *nodes and data---does not work with RBM.
>        So, I think the metadata section is necessary in case of RBM.

Yep, I wouldn't expect an RBM device to use the record_t format at all.  CBJournal would probably use it, though.

> 
>     2. Raw disk layout:
>        Currently, both CBjournal and RBM use record_t to store data (extents and deltas). This is reasonable in the case of CBjournal, but I’m not sure this if RBM is used due to its design characteristic (in-place)---record_t requires additional header when storing data. It seems that we may consider a conventional disk layout for in-place---ondoe contains extents, which indicates where the data is located---like ext3 in the future.
> 
> 
> Any comments?

athanatos commented:
Also, have a look at https://github.com/ceph/ceph/pull/43181.  In addition to supporting a single RBM configuration, it'll be important for this work to end up providing the required machinery for a fast tier for use in concert with a zns device, so it'll need to work with the ExtentPlacementManager multi-device support taking shape in https://github.com/ceph/ceph/pull/43181.

myoungwon commented:
> I'll try to have a look at this this week or next after #43181 merges.
> 
> > @athanatos Before continuing this work, It would be better to get feedback on these commits I have updated.
> > ```
> > 1. I/O path: To make RBM be compatible with current seastore, I think that new extents including small and large block size need to be directly written to RBM, and all deltas excluding large data block need to be appended to CircularBoundedJournal(CBJournal). Then,
> >    after *Cleaner---not implemented yet---reads records on CBJournal, it writes those records in RBM.
> > ```
> 
> I'd like the ExtentPlacementManager out-of-line extent support to be the machinery for this. Have you looked into that yet?

As you can see https://github.com/ceph/ceph/pull/42270/commits/3ce28baa04edb348fd3f9a3932044d988cf2bc1e, I have updated the write strategy on top of EPM. Specifically, new extents (fresh_block_list) excluding small data blocks---In case of new extents, we should reserve the space where the new extents is allocated in RBM area before writing any extent---are added to delayed_alloc_list, which will be directly allocated to the RBM (out-of-line). Otherwise, all deltas (mutated_block_list) are appended to the CBJournal first. 

Note that small data blocks in fresh_block_list also need to be added to CBJournal using delta_info_t, not extent_t (this is not implemented yet). 

> 
> > ```
> > 2. *Cleaner: This is needed to apply deltas from CBJournal to RBM, but this is likely to be the next PR.
> > ```
> 
> I think this probably needs to be a function of the segment cleaner as it's pretty similar to journal trimming with the existing implementation.
> 

Right, this can be inherited from segment cleaner.

> > ```
> > 3. paddr_t: current paddr_t has segment and offset fields. But, RBM probably needs to have block_id (block aligned) which is larger than segment, and block_offset which is less than offset.  So, I have adjusted the size of segment and offset field.
> > ```
> 
> How large is a block? Wouldn't you basically just want a byte offset with an assumption that all extents will be block aligned?
> 

Current segment_id is 32bit, which means it can represent 4 * 4096 GB (16,384) if 4096 bytes block size is used (max offset is 4096).
However, Samsung's recent NVMe SSD (PM1733) already can provide a capacity of 15 TB.


> > The followings are concerns I currently have.
> > ```
> > 1. Metadata section in RBM:
> >    As you know, in RBM, data can be stored randomly across the entire RBM. Therefore, the existing replay mechanism---replaying records by reading record_t sequentially to load *nodes and data---does not work with RBM.
> >    So, I think the metadata section is necessary in case of RBM.
> > ```
> 
> Yep, I wouldn't expect an RBM device to use the record_t format at all. CBJournal would probably use it, though.
> 

Ok.

> > ```
> > 2. Raw disk layout:
> >    Currently, both CBjournal and RBM use record_t to store data (extents and deltas). This is reasonable in the case of CBjournal, but I’m not sure this if RBM is used due to its design characteristic (in-place)---record_t requires additional header when storing data. It seems that we may consider a conventional disk layout for in-place---ondoe contains extents, which indicates where the data is located---like ext3 in the future.
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Any comments?

myoungwon commented:
> Also, have a look at #43181. In addition to supporting a single RBM configuration, it'll be important for this work to end up providing the required machinery for a fast tier for use in concert with a zns device, so it'll need to work with the ExtentPlacementManager multi-device support taking shape in #43181.

I'll review that PR and leave comments from perspective of RBM. But, it would be better if we can proceed these jobs in parallel (discussing what this PR address).

myoungwon commented:
@athanatos I have rebased this PR on top of recent changes. I think this needs to be discussed before any further changes. Can you take a look?

xxhdx1985126 commented on src/crimson/os/seastore/transaction_manager.cc:
```diff
@@ -272,7 +276,8 @@ TransactionManager::submit_transaction_direct(
 
-    return journal->submit_record(std::move(record), tref.get_handle()
+    return jm->submit_record(std::move(record), tref.get_handle(),
+      segment_manager.get_device_id()
```
IIUC, this pr only uses SegmentedJournal, CBJournal hasn't been involved in journaling, right?

xxhdx1985126 commented on src/crimson/os/seastore/extent_placement_manager.h:
```diff
@@ -290,6 +296,97 @@ class SegmentedAllocator : public ExtentAllocator {
+      return trans_intr::do_for_each(alloc_map, [&t](auto& p) {
+        auto writer = p.first;
+        auto& extents_to_persist = p.second;
+	return writer->write(t, extents_to_persist);
```
I think, for random block devices, we don't have to do actual ool writes at the time of delayed extent allocation. The reason why we have to do actual ool writes for segment devices is that, for segment devices, we can only confirm the paddr_t that an extent will be written to after that extent is written. Random block devices don't have this issue, right?

I think maybe we should add a new method like "do_ool_writes()" to EPM, and postpone random block devices' actual ool writes(the invocation of do_ool_writes) to the journal commit phase. With this approach, we can avoid the write amplifications caused by transaction invalidation that happened after the delayed extent allocation. However, this comes at the price of reduced concurrency of ool writes, as only one transaction's ool writes can be done at any time.

@myoungwon @athanatos What do you think?:-)

xxhdx1985126 replied:
> But, in case of a large block (large size record), according to RBM design, we don't have to write the large block to the segment device (CBJournal here), rather we want to write the large block to the RBM directly.

Um, I meant that we just postpone the ool write to the journal commit phase, but still write to RBM instead of CBJournal. In other words, maybe we should give journals the ability to write to other segments/RBMs.

myoungwon replied:
I got it. The postpone is ok. But, I'm not sure that current journal needs to take a role of writing both segment and RBM. Maybe we need a new absraction that can cover both Segment and RBM?

I have one question. When the transaction invalidation happens? Even the invalidation occurs, the write procedure---write for RBM should be preceded before appending record to CBjournal---would not be changes. The invalidation can happen between RBM write and CBjournal write?

xxhdx1985126 replied:
> Maybe we need a new absraction that can cover both Segment and RBM?

Yes, that's also my belief:-)

> I have one question. When the transaction invalidation happens? Even the invalidation occurs, the write procedure---write for RBM should be preceded before appending record to CBjournal---would not be changes. The invalidation can happen between RBM write and CBjournal write?

I think a transaction can be invalidated at any time before it enters the `WritePipeline::prepare` phase, so as long as we do write for RBM after entering that phase, there won't be invalidation between RBM write and CBjournal write.

@athanatos Am I right?:-)

athanatos replied:
Sorry, I've been a bit busy, I'll try to catch up on this tomorrow.

athanatos commented on src/crimson/os/seastore/seastore_types.h:
```diff
@@ -1072,25 +1073,38 @@ struct rbm_alloc_delta_t {
+paddr_t convert_blk_paddr_to_paddr(blk_paddr_t addr, size_t block_size,
+    uint32_t blocks_per_segment, device_id_t d_id);
+blk_paddr_t convert_paddr_to_blk_paddr(paddr_t addr, size_t block_size,
+	uint32_t blocks_per_segment);
```
I see usages of blk_paddr_t in cache.cc and extent_placement_manager.cc.  I'd like to see the RandomBlockDevice interface converted to use paddr_t and all usages of blk_paddr_t restricted to the internals of RandomBlockDevice implementations, as with device_segment_id_t and SegmentManager implementations.

This may mean adjusting the paddr_t representation to remove the explicit segment component.

athanatos commented on src/crimson/os/seastore/random_block_manager/nvme_manager.h:
```diff
@@ -70,6 +72,8 @@ struct rbm_metadata_header_t {
     denc(v.start_data_area, p);
     denc(v.flag, p);
     denc(v.feature, p);
+    denc(v.blocks_per_segment, p);
```
Can we remove the concept of segment from the RandomBlockManager interface and implementations entirely?

athanatos commented on src/crimson/os/seastore/cache.cc:
```diff
@@ -947,6 +947,16 @@ record_t Cache::prepare_record(Transaction &t)
+      blk_paddr_t rbm_addr =
+	convert_paddr_to_blk_paddr(paddr, rbm->get_block_size(),
+	  rbm->get_blocks_per_segment());
+      rbm->free_extent(t,
```
This appears to be the only user of the rpm member.  I don't think this is a good place for this.  In general, there may be more than one RandomBlockDevice device.  Every other read in Cache deals with that possibility by routing the read through the ExtentReader interfaces.

What does free_extent do here?  NVMeManager::add_free_extent seem to just add an rbm_alloc_delta_t to Transaction::rbm_alloc_info_blocks.  The only thing the NVMeManager instance is actually used for is to translate the blk_paddr_t to a blk_id_t.  This process seems as though it *must* be identical for every RandomBlockManger implementation since it's constrained by the journal format and Transaction structures.  Thus:
1) Change rbm_alloc_delta_t to use an interval_set<rbm_offset_t> determined directly from the paddr_t.
2) Remove RandomBlockManager::free_extent and Cache::rbm entirely -- every implementation must behave the same way.

athanatos commented on src/crimson/os/seastore/random_block_manager/nvme_manager.cc:
```diff
@@ -253,20 +255,21 @@ NVMeManager::allocate_ertr::future<> NVMeManager::alloc_extent(
 	    rbm_alloc_delta_t::op_types_t::SET
 	  };
-	  // TODO: add alloc info to delta
+	  t.add_rbm_alloc_info_blocks(alloc_info);
```
I think I just forgot this from before.  How do we ensure that multiple concurrent transactions don't allocate the same block (or detect the conflict if they do?)

athanatos commented on src/crimson/os/seastore/random_block_manager.h:
```diff
@@ -33,6 +33,7 @@ class RandomBlockManager {
     size_t block_size = 0;
```
Fix 'RBM: add device_id_t' commit message.
